// @ngneat/error-tailor/esm5/lib/control-error.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, TemplateRef } from '@angular/core';
var DefaultControlErrorComponent = /** @class */ (function () {
    function DefaultControlErrorComponent(cdr, host) {
        this.cdr = cdr;
        this.host = host;
        this.errorText = null;
        this.hideError = true;
    }
    DefaultControlErrorComponent.prototype.createTemplate = function (tpl, error, text) {
        this.errorTemplate = tpl;
        this.errorContext = { $implicit: error, text: text };
        this.cdr.markForCheck();
    };
    Object.defineProperty(DefaultControlErrorComponent.prototype, "customClass", {
        set: function (className) {
            this.host.nativeElement.classList.add(className);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DefaultControlErrorComponent.prototype, "text", {
        set: function (value) {
            if (value !== this.errorText) {
                this.errorText = value;
                this.hideError = !value;
                this.cdr.markForCheck();
            }
        },
        enumerable: true,
        configurable: true
    });
    DefaultControlErrorComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ElementRef }
    ]; };
    DefaultControlErrorComponent = __decorate([
        Component({
            selector: 'control-error',
            template: "\n    <label class=\"control-error\" [class.hide-control]=\"hideError\" *ngIf=\"!errorTemplate\">{{ errorText }}</label>\n    <ng-template *ngTemplateOutlet=\"errorTemplate; context: errorContext\"></ng-template>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["\n      .hide-control {\n        display: none !important;\n      }\n\n      :host {\n        display: block;\n      }\n    "]
        })
    ], DefaultControlErrorComponent);
    return DefaultControlErrorComponent;
}());
export { DefaultControlErrorComponent };
