// @ngneat/error-tailor/esm5/lib/error-tailor.module.js
import { __assign, __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { ControlErrorsDirective } from './control-error.directive';
import { ControlErrorAnchorDirective } from './control-error-anchor.directive';
import { DefaultControlErrorComponent } from './control-error.component';
import { CommonModule } from '@angular/common';
import { FormActionDirective } from './form-action.directive';
import { ErrorTailorConfigProvider, FORM_ERRORS } from './providers';
var api = [DefaultControlErrorComponent, ControlErrorAnchorDirective, ControlErrorsDirective, FormActionDirective];
var ErrorTailorModule = /** @class */ (function () {
    function ErrorTailorModule() {
    }
    ErrorTailorModule_1 = ErrorTailorModule;
    ErrorTailorModule.forRoot = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: ErrorTailorModule_1,
            providers: [
                {
                    provide: ErrorTailorConfigProvider,
                    useValue: config
                },
                __assign({ provide: FORM_ERRORS }, config.errors)
            ]
        };
    };
    var ErrorTailorModule_1;
    ErrorTailorModule = ErrorTailorModule_1 = __decorate([
        NgModule({
            declarations: [
                ControlErrorsDirective,
                ControlErrorAnchorDirective,
                DefaultControlErrorComponent,
                FormActionDirective
            ],
            imports: [CommonModule],
            exports: [api],
            entryComponents: [DefaultControlErrorComponent]
        })
    ], ErrorTailorModule);
    return ErrorTailorModule;
}());
export { ErrorTailorModule };
