// @ngneat/error-tailor/bundles/ngneat-error-tailor.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('rxjs'), require('rxjs/operators'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@ngneat/error-tailor', ['exports', '@angular/core', '@angular/forms', 'rxjs', 'rxjs/operators', '@angular/common'], factory) :
    (global = global || self, factory((global.ngneat = global.ngneat || {}, global.ngneat['error-tailor'] = {}), global.ng.core, global.ng.forms, global.rxjs, global.rxjs.operators, global.ng.common));
}(this, (function (exports, core, forms, rxjs, operators, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var DefaultControlErrorComponent = /** @class */ (function () {
        function DefaultControlErrorComponent(cdr, host) {
            this.cdr = cdr;
            this.host = host;
            this.errorText = null;
            this.hideError = true;
        }
        DefaultControlErrorComponent.prototype.createTemplate = function (tpl, error, text) {
            this.errorTemplate = tpl;
            this.errorContext = { $implicit: error, text: text };
            this.cdr.markForCheck();
        };
        Object.defineProperty(DefaultControlErrorComponent.prototype, "customClass", {
            set: function (className) {
                this.host.nativeElement.classList.add(className);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DefaultControlErrorComponent.prototype, "text", {
            set: function (value) {
                if (value !== this.errorText) {
                    this.errorText = value;
                    this.hideError = !value;
                    this.cdr.markForCheck();
                }
            },
            enumerable: true,
            configurable: true
        });
        DefaultControlErrorComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef },
            { type: core.ElementRef }
        ]; };
        DefaultControlErrorComponent = __decorate([
            core.Component({
                selector: 'control-error',
                template: "\n    <label class=\"control-error\" [class.hide-control]=\"hideError\" *ngIf=\"!errorTemplate\">{{ errorText }}</label>\n    <ng-template *ngTemplateOutlet=\"errorTemplate; context: errorContext\"></ng-template>\n  ",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: ["\n      .hide-control {\n        display: none !important;\n      }\n\n      :host {\n        display: block;\n      }\n    "]
            })
        ], DefaultControlErrorComponent);
        return DefaultControlErrorComponent;
    }());

    var ControlErrorAnchorDirective = /** @class */ (function () {
        function ControlErrorAnchorDirective(vcr) {
            this.vcr = vcr;
        }
        ControlErrorAnchorDirective.ctorParameters = function () { return [
            { type: core.ViewContainerRef }
        ]; };
        ControlErrorAnchorDirective = __decorate([
            core.Directive({
                selector: '[controlErrorAnchor]',
                exportAs: 'controlErrorAnchor'
            })
        ], ControlErrorAnchorDirective);
        return ControlErrorAnchorDirective;
    }());

    var FORM_ERRORS = new core.InjectionToken('FORM_ERRORS', {
        providedIn: 'root',
        factory: function () {
            return {};
        }
    });
    var ErrorTailorConfigProvider = new core.InjectionToken('ErrorTailorConfigProvider');

    var FormActionDirective = /** @class */ (function () {
        function FormActionDirective(host) {
            var _this = this;
            this.host = host;
            this.submit$ = rxjs.fromEvent(this.element, 'submit').pipe(operators.tap(function () {
                if (_this.element.classList.contains('form-submitted') === false) {
                    _this.element.classList.add('form-submitted');
                }
            }), operators.shareReplay({ refCount: true, bufferSize: 1 }));
            this.reset$ = rxjs.fromEvent(this.element, 'reset').pipe(operators.tap(function () {
                _this.element.classList.remove('form-submitted');
            }), operators.shareReplay({ refCount: true, bufferSize: 1 }));
        }
        Object.defineProperty(FormActionDirective.prototype, "element", {
            get: function () {
                return this.host.nativeElement;
            },
            enumerable: true,
            configurable: true
        });
        FormActionDirective.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        FormActionDirective = __decorate([
            core.Directive({
                selector: 'form[errorTailor]'
            })
        ], FormActionDirective);
        return FormActionDirective;
    }());

    var ControlErrorsDirective = /** @class */ (function () {
        function ControlErrorsDirective(vcr, resolver, host, config, globalErrors, controlErrorAnchorParent, form, ngControl, controlContainer) {
            this.vcr = vcr;
            this.resolver = resolver;
            this.host = host;
            this.config = config;
            this.globalErrors = globalErrors;
            this.controlErrorAnchorParent = controlErrorAnchorParent;
            this.form = form;
            this.ngControl = ngControl;
            this.controlContainer = controlContainer;
            this.customErrors = {};
            this.controlErrorsOnAsync = true;
            this.controlErrorsOnBlur = true;
            this.destroy = new rxjs.Subject();
            this.mergedConfig = {};
            this.submit$ = this.form ? this.form.submit$ : rxjs.EMPTY;
            this.reset$ = this.form ? this.form.reset$ : rxjs.EMPTY;
            this.mergedConfig = this.buildConfig();
        }
        ControlErrorsDirective.prototype.ngOnInit = function () {
            var _this = this;
            this.anchor = this.resolveAnchor();
            this.control = (this.controlContainer || this.ngControl).control;
            var hasAsyncValidator = !!this.control.asyncValidator;
            var isInput = this.mergedConfig.blurPredicate(this.host.nativeElement);
            var statusChanges$ = this.control.statusChanges.pipe(operators.distinctUntilChanged());
            var valueChanges$ = this.control.valueChanges;
            var controlChanges$ = rxjs.merge(statusChanges$, valueChanges$);
            var changesOnAsync$ = rxjs.EMPTY;
            var changesOnBlur$ = rxjs.EMPTY;
            if (this.controlErrorsOnAsync && hasAsyncValidator) {
                // hasAsyncThenUponStatusChange
                changesOnAsync$ = statusChanges$.pipe(operators.startWith(true));
            }
            if (this.controlErrorsOnBlur && isInput) {
                var blur$ = rxjs.fromEvent(this.host.nativeElement, 'focusout');
                // blurFirstThenUponChange
                changesOnBlur$ = blur$.pipe(operators.switchMap(function () { return valueChanges$.pipe(operators.startWith(true)); }));
            }
            var submit$ = rxjs.merge(this.submit$.pipe(operators.mapTo(true)), this.reset$.pipe(operators.mapTo(false)));
            // when submitted, submitFirstThenUponChanges
            var changesOnSubmit$ = submit$.pipe(operators.switchMap(function (submit) { return (submit ? controlChanges$.pipe(operators.startWith(true)) : rxjs.NEVER); }));
            // on reset, clear ComponentRef and customAnchorDestroyFn
            this.reset$.pipe(operators.takeUntil(this.destroy)).subscribe(function () { return _this.clearRefs(); });
            rxjs.merge(changesOnAsync$, changesOnBlur$, changesOnSubmit$)
                .pipe(operators.takeUntil(this.destroy))
                .subscribe(function () { return _this.valueChanges(); });
        };
        ControlErrorsDirective.prototype.setError = function (text, error) {
            if (!this.ref) {
                var factory = this.resolver.resolveComponentFactory(this.mergedConfig.controlErrorComponent);
                this.ref = this.anchor.createComponent(factory);
            }
            var instance = this.ref.instance;
            if (this.controlErrorsTpl) {
                instance.createTemplate(this.controlErrorsTpl, error, text);
            }
            else {
                instance.text = text;
            }
            if (this.controlErrorsClass) {
                instance.customClass = this.controlErrorsClass;
            }
            if (this.mergedConfig.controlErrorComponentAnchorFn) {
                this.customAnchorDestroyFn = this.mergedConfig.controlErrorComponentAnchorFn(this.host.nativeElement, this.ref.hostView.rootNodes[0]);
            }
        };
        ControlErrorsDirective.prototype.ngOnDestroy = function () {
            this.destroy.next();
            this.clearRefs();
        };
        ControlErrorsDirective.prototype.clearRefs = function () {
            if (this.customAnchorDestroyFn) {
                this.customAnchorDestroyFn();
                this.customAnchorDestroyFn = null;
            }
            if (this.ref) {
                this.ref.destroy();
            }
            this.ref = null;
        };
        ControlErrorsDirective.prototype.valueChanges = function () {
            var controlErrors = this.control.errors;
            if (controlErrors) {
                var _a = __read(Object.keys(controlErrors), 1), firstKey = _a[0];
                var getError = this.customErrors[firstKey] || this.globalErrors[firstKey];
                if (!getError) {
                    return;
                }
                var text = typeof getError === 'function' ? getError(controlErrors[firstKey]) : getError;
                this.setError(text, controlErrors);
            }
            else if (this.ref) {
                this.setError(null);
            }
        };
        ControlErrorsDirective.prototype.resolveAnchor = function () {
            if (this.controlErrorAnchor) {
                return this.controlErrorAnchor.vcr;
            }
            if (this.controlErrorAnchorParent) {
                return this.controlErrorAnchorParent.vcr;
            }
            return this.vcr;
        };
        ControlErrorsDirective.prototype.buildConfig = function () {
            return __assign({
                blurPredicate: function (element) {
                    return element.tagName === 'INPUT' || element.tagName === 'SELECT';
                },
                controlErrorComponent: DefaultControlErrorComponent
            }, this.config);
        };
        ControlErrorsDirective.ctorParameters = function () { return [
            { type: core.ViewContainerRef },
            { type: core.ComponentFactoryResolver },
            { type: core.ElementRef },
            { type: undefined, decorators: [{ type: core.Inject, args: [ErrorTailorConfigProvider,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [FORM_ERRORS,] }] },
            { type: ControlErrorAnchorDirective, decorators: [{ type: core.Optional }] },
            { type: FormActionDirective, decorators: [{ type: core.Optional }] },
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }] },
            { type: forms.ControlContainer, decorators: [{ type: core.Optional }, { type: core.Self }] }
        ]; };
        __decorate([
            core.Input('controlErrors')
        ], ControlErrorsDirective.prototype, "customErrors", void 0);
        __decorate([
            core.Input()
        ], ControlErrorsDirective.prototype, "controlErrorsClass", void 0);
        __decorate([
            core.Input()
        ], ControlErrorsDirective.prototype, "controlErrorsTpl", void 0);
        __decorate([
            core.Input()
        ], ControlErrorsDirective.prototype, "controlErrorsOnAsync", void 0);
        __decorate([
            core.Input()
        ], ControlErrorsDirective.prototype, "controlErrorsOnBlur", void 0);
        __decorate([
            core.Input()
        ], ControlErrorsDirective.prototype, "controlErrorAnchor", void 0);
        ControlErrorsDirective = __decorate([
            core.Directive({
                selector: '[formControlName]:not([controlErrorsIgnore]), [formControl]:not([controlErrorsIgnore]), [formGroup]:not([controlErrorsIgnore]), [formGroupName]:not([controlErrorsIgnore]), [formArrayName]:not([controlErrorsIgnore]), [ngModel]:not([controlErrorsIgnore])'
            }),
            __param(3, core.Inject(ErrorTailorConfigProvider)),
            __param(4, core.Inject(FORM_ERRORS)),
            __param(5, core.Optional()),
            __param(6, core.Optional()),
            __param(7, core.Optional()), __param(7, core.Self()),
            __param(8, core.Optional()), __param(8, core.Self())
        ], ControlErrorsDirective);
        return ControlErrorsDirective;
    }());

    var api = [DefaultControlErrorComponent, ControlErrorAnchorDirective, ControlErrorsDirective, FormActionDirective];
    var ErrorTailorModule = /** @class */ (function () {
        function ErrorTailorModule() {
        }
        ErrorTailorModule_1 = ErrorTailorModule;
        ErrorTailorModule.forRoot = function (config) {
            if (config === void 0) { config = {}; }
            return {
                ngModule: ErrorTailorModule_1,
                providers: [
                    {
                        provide: ErrorTailorConfigProvider,
                        useValue: config
                    },
                    __assign({ provide: FORM_ERRORS }, config.errors)
                ]
            };
        };
        var ErrorTailorModule_1;
        ErrorTailorModule = ErrorTailorModule_1 = __decorate([
            core.NgModule({
                declarations: [
                    ControlErrorsDirective,
                    ControlErrorAnchorDirective,
                    DefaultControlErrorComponent,
                    FormActionDirective
                ],
                imports: [common.CommonModule],
                exports: [api],
                entryComponents: [DefaultControlErrorComponent]
            })
        ], ErrorTailorModule);
        return ErrorTailorModule;
    }());

    exports.ControlErrorAnchorDirective = ControlErrorAnchorDirective;
    exports.ControlErrorsDirective = ControlErrorsDirective;
    exports.DefaultControlErrorComponent = DefaultControlErrorComponent;
    exports.ErrorTailorConfigProvider = ErrorTailorConfigProvider;
    exports.ErrorTailorModule = ErrorTailorModule;
    exports.FORM_ERRORS = FORM_ERRORS;
    exports.FormActionDirective = FormActionDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
