// @ngneat/error-tailor/esm2015/lib/control-error.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, TemplateRef } from '@angular/core';
let DefaultControlErrorComponent = class DefaultControlErrorComponent {
    constructor(cdr, host) {
        this.cdr = cdr;
        this.host = host;
        this.errorText = null;
        this.hideError = true;
    }
    createTemplate(tpl, error, text) {
        this.errorTemplate = tpl;
        this.errorContext = { $implicit: error, text };
        this.cdr.markForCheck();
    }
    set customClass(className) {
        this.host.nativeElement.classList.add(className);
    }
    set text(value) {
        if (value !== this.errorText) {
            this.errorText = value;
            this.hideError = !value;
            this.cdr.markForCheck();
        }
    }
};
DefaultControlErrorComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef }
];
DefaultControlErrorComponent = __decorate([
    Component({
        selector: 'control-error',
        template: `
    <label class="control-error" [class.hide-control]="hideError" *ngIf="!errorTemplate">{{ errorText }}</label>
    <ng-template *ngTemplateOutlet="errorTemplate; context: errorContext"></ng-template>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [`
      .hide-control {
        display: none !important;
      }

      :host {
        display: block;
      }
    `]
    })
], DefaultControlErrorComponent);
export { DefaultControlErrorComponent };
