// @ngneat/error-tailor/esm2015/lib/form-action.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef } from '@angular/core';
import { fromEvent } from 'rxjs';
import { shareReplay, tap } from 'rxjs/operators';
let FormActionDirective = class FormActionDirective {
    constructor(host) {
        this.host = host;
        this.submit$ = fromEvent(this.element, 'submit').pipe(tap(() => {
            if (this.element.classList.contains('form-submitted') === false) {
                this.element.classList.add('form-submitted');
            }
        }), shareReplay({ refCount: true, bufferSize: 1 }));
        this.reset$ = fromEvent(this.element, 'reset').pipe(tap(() => {
            this.element.classList.remove('form-submitted');
        }), shareReplay({ refCount: true, bufferSize: 1 }));
    }
    get element() {
        return this.host.nativeElement;
    }
};
FormActionDirective.ctorParameters = () => [
    { type: ElementRef }
];
FormActionDirective = __decorate([
    Directive({
        selector: 'form[errorTailor]'
    })
], FormActionDirective);
export { FormActionDirective };
